glean-core/rlb/tests/schema.rs

```diff
@@ -48,39 +52,63 @@ fn new_glean() -> tempfile::TempDir {
 
-    let dir = new_glean();
-
-    // Register and submit a ping for testing
-    let ping_type = PingType::new("test", true, /* send_if_empty */ true, vec![]);
-    glean::register_ping_type(&ping_type);
-    ping_type.submit(None);
-    glean::dispatcher::block_on_queue();
-
-    // Read the ping from disk.
-    // We know where it should be placed.
-    let pings_dir = dir.path().join("pending_pings");
-    let pending_pings: Vec<PathBuf> = pings_dir
-        .read_dir()
-        .unwrap()
-        .filter_map(|entry| entry.ok())
-        .filter_map(|entry| match entry.file_type() {
-            Ok(f) if f.is_file() => Some(entry.path()),
-            _ => None,
-        })
-        .collect();
-
-    // There should only be one: our custom ping.
-    assert_eq!(1, pending_pings.len());
-
-    // 1. line: the URL
-    // 2. line: the JSON body
-    let file = File::open(&pending_pings[0]).unwrap();
-    let mut lines = BufReader::new(file).lines();
-    let _url = lines.next().unwrap().unwrap();
-    let body = lines.next().unwrap().unwrap();
+    let (s, r) = crossbeam_channel::bounded::<Vec<u8>>(1);
+
+    // Define a fake uploader that reports back the submission URL
+    // using a crossbeam channel.
+    #[derive(Debug)]
+    pub struct ValidatingUploader {
+        sender: crossbeam_channel::Sender<Vec<u8>>,
+    };
+    impl glean::net::PingUploader for ValidatingUploader {
+        fn upload(
+            &self,
+            _url: String,
+            body: Vec<u8>,
+            _headers: Vec<(String, String)>,
+        ) -> glean::net::UploadResult {
+            self.sender.send(body).unwrap();
+            glean::net::UploadResult::HttpStatus(200)
+        }
+    }
+
+    // Create a custom configuration to use a validating uploader.
+    let dir = tempfile::tempdir().unwrap();
+    let tmpname = dir.path().display().to_string();
+
+    let cfg = Configuration {
+        data_path: tmpname,
+        application_id: GLOBAL_APPLICATION_ID.into(),
+        upload_enabled: true,
+        max_events: None,
+        delay_ping_lifetime_io: false,
+        channel: Some("testing".into()),
+        server_endpoint: Some("invalid-test-host".into()),
+        uploader: Some(Box::new(ValidatingUploader { sender: s })),
+    };
+    let _ = new_glean(Some(cfg));
+
+    // Define a new ping and submit it.
+    const PING_NAME: &str = "test-ping";
+    let custom_ping = glean::private::PingType::new(PING_NAME, true, true, vec![]);
+    custom_ping.submit(None);
+
+    // Wait for the ping to arrive.
+    let raw_body = r.recv().unwrap();
+
+    // Decode the gzipped body.
+    let mut gzip_decoder = GzDecoder::new(&raw_body[..]);
+    let mut s = String::with_capacity(raw_body.len());
+
+    let data = gzip_decoder
+        .read_to_string(&mut s)
+        .ok()
+        .map(|_| &s[..])
+        .or_else(|| std::str::from_utf8(&raw_body).ok())
+        .and_then(|payload| serde_json::from_str(payload).ok())
+        .unwrap();
```

Comment:
Yikes, good opportunity to factor out later.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1675476